DRAFT: next-auth provider profile

usrrname opened the pull request:
This is a PR meant for figuring out what we might need to do to complete this integration.

- [ ] Add the provider logo to the docs/public/img/providers directory.
- [x] Add the provider in the drop-down list in [.github/ISSUE_TEMPLATE/2_bug_provider.yml](https://github.com/nextauthjs/next-auth/blob/main/.github/ISSUE_TEMPLATE/2_bug_provider.yml)

Issue template starts below:

---------

<!--
Thanks for your interest in the project. Bugs filed and PRs submitted are appreciated!

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!

**NOTE**:

- It's a good idea to open an issue first to discuss potential changes.
- Please make sure that you are _NOT_ opening a PR to fix a potential security vulnerability. Instead, please follow the [Security guidelines](https://github.com/nextauthjs/.github/blob/main/SECURITY.md) to disclose the issue to us confidentially.

-->

## ☕️ Reasoning

The purpose of this PR is to add `@curveball/a12n-server` as a built-in provider.
`@curveball/a12n-server` is an open source OAuth2 authentication system that has been maintained since 2018.

<!-- What changes are being made? What feature/bug is being fixed here? -->

## 🧢 Checklist

- [ ] Documentation
- [ ] Tests
- [ ] Ready to be merged

## 🎫 Affected issues
https://github.com/curveball/a12n-server/issues/585 
<!--
Please [scout and link issues](https://github.com/nextauthjs/next-auth/issues) that might be solved by this PR. And include text like the following to close them automatically when this is merged:

Fixes: INSERT_ISSUE_LINK_HERE
-->

## 📌 Resources

- [Security guidelines](https://github.com/nextauthjs/.github/blob/main/SECURITY.md)
- [Contributing guidelines](https://github.com/nextauthjs/.github/blob/main/CONTRIBUTING.md)
- [Code of conduct](https://github.com/nextauthjs/.github/blob/main/CODE_OF_CONDUCT.md)
- [Contributing to Open Source](https://kcd.im/pull-request)

Commit:
initial commit

Commit:
initial commit

Commit:
initial commit

usrrname commented on packages/core/src/providers/a12n-server.ts:
```diff
@@ -0,0 +1,111 @@
+/**
+ * @module providers/a12n-server
+ */
+export interface A12nServerProfile extends Record<string, any> {
```
This type was copied from IntrospectionResponse

evert replied:
I think we might need to use the format from `userinfo` instead of the format from the introspection endpoint, which as of right now is a lot simpler:

```
type UserInfo = {
  sub: string;
  email?: string;
  email_verified?: boolean;
  name: string;
  website?: string;
  zoneinfo?: string;
  given_name?: string;
  family_name?: string;
  preferred_username?: string;
  phone_number?: string;
  phone_number_verified?: boolean;
  locale?: string;
  updated_at: number;
  picture?: string;
  address?: string;
  birthdate?: string;
}
```

But I'm not a 100% sure of this!

usrrname replied:
I chose the IntrospectResponse type as it had a lot of what the other example's responses such as: 
iss, sub, iat, jti... and I realize I was looking at oidc type providers.

If we are going with oauth type,  I think some of the user info could be part of a [profile](https://github.com/curveball/next-auth/blob/31ead6df29c508b8177f6e088cd4f11f3a341886/packages/core/src/providers/atlassian.ts#L108) callback function on the config object ...

1. "oauth" type provider examples:

- atlassian: https://github.com/curveball/next-auth/blob/main/packages/core/src/providers/atlassian.ts

- mastodon : https://github.com/curveball/next-auth/blob/main/packages/core/src/providers/mastodon.ts

`next-auth/blob/main/packages/core/src/providers/oauth.ts` seems to indicate that we may have to adapt the response to `Callback<Profile>`: https://github.com/curveball/next-auth/blob/main/packages/core/src/providers/oauth.ts#L147-L154 

2. "oidc" type providers:

- asgardeo https://github.com/curveball/next-auth/blob/main/packages/core/src/providers/asgardeo.ts
-  twitch https://github.com/curveball/next-auth/blob/main/packages/core/src/providers/twitch.ts 

Which type should we go with? My latest edit stuck with oauth

usrrname replied:
OHH! next-auth's `Profile` type corresponds with https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims , and so does all the `UserInfo` you mentioned here https://github.com/curveball/next-auth/pull/1#discussion_r1938554961

usrrname commented on packages/core/src/providers/a12n-server.ts:
```diff
@@ -0,0 +1,111 @@
+ * - Sign in with a12n-server [REST API]()
+ * - [How to retrieve]() the user's information from your a12n-server
+ * - [Learn more about OAuth](https://authjs.dev/concepts/oauth)
+ * - [Creating the Client Secret]()
```
Todo: docs

evert replied:
This may be helpful, these urls take you through the flow of setting up a client automatically:


```
http://localhost:8531/app/new?nickname=nextjs-a12n&allowedGrantTypes=authorization_code,refresh_token&redirectUris=http://localhost:3000/callback.html&url=http://localhost:3000&clientId=nextjs-a12n
```

usrrname replied:
I can't tell if admins should be able to create and see their client_id and client secret in the dashboard or if I'm imagining a different kind of dev experience 🤔

usrrname replied:
Nevermind, I see where to find it now in a12n-server...

evert commented on packages/core/src/providers/a12n-server.ts:
```diff
@@ -0,0 +1,111 @@
+        issuer: config.issuer,
+        clientId: config.clientId, 
+        clientSecret: config.clientSecret, 
+        wellKnown: config.issuer + "/.well-known/oauth-authorization-server",
```
I think this one is no longer needed, because it will now be correctly inferred from `issuer`

usrrname replied:
```suggestion
```

usrrname committed:
chore: remove wellKnown property: issuer will indicate meta

usrrname committed:
chore: remove wellKnown property: issuer will indicate meta

usrrname committed:
chore: remove wellKnown property: issuer will indicate meta

usrrname committed:
refactor: there was an attempt to use existing types from authjs

usrrname committed:
refactor: there was an attempt to use existing types from authjs

usrrname committed:
refactor: there was an attempt to use existing types from authjs

usrrname committed:
refactor: change to oidc config

usrrname committed:
refactor: change to oidc config

usrrname committed:
refactor: change to oidc config

usrrname committed:
feat: update env var with prefix of AUTH_

usrrname committed:
feat: update env var with prefix of AUTH_

usrrname committed:
feat: update env var with prefix of AUTH_

usrrname commented on packages/core/src/providers/a12n-server.ts:
```diff
@@ -0,0 +1,126 @@
+ *
+ * - a12n-server [Overview](https://github.com/curveball/a12n-server/blob/main/docs/getting-started.md)
+ * - Sign in with a12n-server [REST API]()
+ * - [How to add a new client to a12n-server]()
```
```suggestion
 * - [How to add a new client to a12n-server](https://github.com/curveball/next-a12n/tree/main?tab=readme-ov-file#register-a-new-client-side-web-app-on-curveballa12n-server)
```

usrrname committed:
Update packages/core/src/providers/a12n-server.ts

chore: adds a12n-server to provider dropdown

chore: update env variable format

doc: update doc in comments

usrrname committed:
Update packages/core/src/providers/a12n-server.ts

usrrname committed:
Update packages/core/src/providers/a12n-server.ts

Commit:
chore: adds a12n-server to provider dropdown

Commit:
chore: adds a12n-server to provider dropdown

Commit:
chore: update env variable format

Commit:
doc: update doc in comments

usrrname committed:
Merge branch 'feat/add-a12n-server' of github.com:curveball/next-auth into feat/add-a12n-server

usrrname committed:
Merge branch 'feat/add-a12n-server' of github.com:curveball/next-auth into feat/add-a12n-server
